Lib9c/Action/ItemEnhancement4.cs

```diff
@@ -0,0 +1,254 @@
+                );
+            }
+
+            if (!avatarState.inventory.TryGetNonFungibleItem(materialId, out ItemUsable materialItem))
+            {
+                throw new NotEnoughMaterialException(
+                    $"Aborted as the signer does not have a necessary material ({materialId})."
+                );
+            }
+
+            if (materialItem.RequiredBlockIndex > context.BlockIndex)
+            {
+                throw new RequiredBlockIndexException(
+                    $"Aborted as the material ({materialId}) is not available yet; it will be available at the block #{materialItem.RequiredBlockIndex}."
+                );
+            }
+
+            if (!(materialItem is Equipment materialEquipment))
+            {
+                throw new InvalidCastException(
+                    $"Aborted as the material item is not an {nameof(Equipment)}, but {materialItem.GetType().Name}."
+                );
+            }
+
+            if (enhancementEquipment.ItemId == materialId)
+            {
+                throw new InvalidMaterialException(
+                    $"Aborted as an equipment to enhance ({materialId}) was used as a material too."
+                );
+            }
+
+            if (materialEquipment.ItemSubType != enhancementEquipment.ItemSubType)
+            {
+                // Invalid ItemSubType
+                throw new InvalidMaterialException(
+                    $"Aborted as the material item is not a {enhancementEquipment.ItemSubType}, but {materialEquipment.ItemSubType}."
+                );
+            }
+
+            if (materialEquipment.Grade != enhancementEquipment.Grade)
+            {
+                // Invalid Grade
+                throw new InvalidMaterialException(
+                    $"Aborted as grades of the equipment to enhance ({enhancementEquipment.Grade}) and a material ({materialEquipment.Grade}) does not match."
+                );
+            }
+
+            if (materialEquipment.level != enhancementEquipment.level)
+            {
+                // Invalid level
+                throw new InvalidMaterialException(
+                    $"Aborted as levels of the equipment to enhance ({enhancementEquipment.level}) and a material ({materialEquipment.level}) does not match."
+                );
+            }
+            sw.Stop();
+            Log.Debug("ItemEnhancement Get Material: {Elapsed}", sw.Elapsed);
+            sw.Restart();
+            materialEquipment.Unequip();
+
+            enhancementEquipment.Unequip();
+
+            enhancementEquipment = ItemEnhancement.UpgradeEquipment(enhancementEquipment);
+
+            var requiredBlockIndex = ctx.BlockIndex + RequiredBlockCount;
+            enhancementEquipment.Update(requiredBlockIndex);
+            sw.Stop();
+            Log.Debug("ItemEnhancement Upgrade Equipment: {Elapsed}", sw.Elapsed);
+            sw.Restart();
+
+            result.gold = requiredNCG;
```

Comment:
```suggestion
            result.gold = 0;
```

This line is changed.